elasticsearch/elasticsearch.go

```diff
@@ -39,20 +39,83 @@ const (
 	ExternalGTE = "external_gte"
 )
 
-var log = loggo.GetLogger("charmstore.elasticsearch")
+var logger = loggo.GetLogger("charmstore.elasticsearch")
 
-var ErrConflict = errgo.New("elasticsearch document conflict")
-var ErrNotFound = errgo.New("elasticsearch document not found")
+// An ElasticsearchError represents an error returned by the
+// elasticsearch server. All methods in this package return errors of
+// this type where the error was return from the elasticsearch service.
+type ElasticsearchError struct {
+	// Status contains the status code returned from the
+	// elasticsearch server.
+	Status int
 
-type ElasticSearchError struct {
-	Err    string `json:"error"`
-	Status int    `json:"status"`
+	// body contains the body of the error response.
+	body errorBody
 }
 
-func (e ElasticSearchError) Error() string {
-	return e.Err
+func (e *ElasticsearchError) Error() string {
+	s := strings.ToLower(http.StatusText(e.Status))
+	if e.body.Error.Reason != "" {
+		s += ": " + e.body.Error.Reason
+	}
+	return s
+}
+
+// Type returns the type specified in the elasticsearch error response,
+// if any.
+func (e *ElasticsearchError) Type() string {
+	return e.body.Error.Type
 }
 
+type errorBody struct {
+	Status int       `json:"status"`
+	Error  errorInfo `json:"error"`
+}
+
+type errorInfo struct {
+	Type   string `json:"type"`
+	Reason string `json:"reason"`
+}
+
+// UnmarshalJSON implements json.Unmarshaler
+func (e *errorInfo) UnmarshalJSON(b []byte) error {
+	type ei errorInfo
+	if b[0] == '{' {
```

Comment:
what if b is empty ?

Reply:
@rogpeppe tells me it won't be. The json package won't call it with an empty b.